[FSDP2] Improved msg when parent mesh but not DTensor

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #120952
* #121357
* #121356
* __->__ #121244
* #121328
* #120351

Currently, FSDP2 requires that if using a `mesh` that has a parent mesh, then the parameters to shard must be `DTensor`s. This PR improves the error message to help find the parameter that violates this.

Do we _have_ to have this requirement? Maybe not. The sharded state dict could still use the 1D mesh. However, I worry that we could inadvertently have a case where the sharded state dict then has a mix of 1D and 2D `DTensor`s.

cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang

Commit:
[FSDP2] Improved msg when parent mesh but not DTensor

[ghstack-poisoned]

wanchaol commented on torch/distributed/_composable/fsdp/_fsdp_param.py:
```diff
@@ -183,11 +183,14 @@ def _init_sharded_param(self, param: nn.Parameter, device: torch.device):
         else:
-            if _mesh_resources.get_parent_mesh(self.mesh_info.mesh) is not None:
+            mesh = self.mesh_info.mesh
+            if (parent_mesh := _mesh_resources.get_parent_mesh(mesh)) is not None:
```
hmmm what if user only apply TP to some submodules, and the other submodules are not DTensors but they are replicated (either manual replication or always replicated like torch.ones param). Would applying FSDP on top of this model fail?

awgu replied:
Today, it would fail. The question to me is what should be the sharded state dict look like, and we can go from there. If it is okay for us to mix 1D and 2D DTensors in the same sharded state dict, then we can simply relax this check instead of erroring.

mvpatel2000 replied:
> If it is okay for us to mix 1D and 2D DTensors in the same sharded state dict

We do this today with FSDP1 and it works great :)

awgu replied:
Nice! Okay, we can probably relax this check then :)

Commit:
Update on "[FSDP2] Improved msg when parent mesh but not DTensor"


Currently, FSDP2 requires that if using a `mesh` that has a parent mesh, then the parameters to shard must be `DTensor`s. This PR improves the error message to help find the parameter that violates this.

Do we _have_ to have this requirement? Maybe not. The sharded state dict could still use the 1D mesh. However, I worry that we could inadvertently have a case where the sharded state dict then has a mix of 1D and 2D `DTensor`s.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang

[ghstack-poisoned]
